merging: lock index dir during merge

stefanhengl opened the pull request:
fixes: #319 

This updates mainly `doMerge`, which drives the merge process. Previously we didn't lock the index dir which can potentially lead to race conditions with indexing and cleanup. 

Changes:
- lock index dir during merging. We release the lock after each compound shard we create
- set default merge interval from 1h to 8h
- recompute candidates before each merge, because candidates might change between two merges
- remove simulation mode
- the debug command is now curl-based and respects the lock

There is a use case of the debug command I haven't thought about before. Customers can use it to trigger an inital merge during less busy hours even if shard merging is disabled.

### Test plan
- new unit tests
- manual testing
  - I ran a local instance of indexserver with shard merging enabled 

  ```
  SRC_ENABLE_SHARD_MERGING=true go run . --merge_target_size 40 --merge_min_size 5 --merge_interval 10s -- 
  sourcegraph_url <local/dir> --debug
  ```

  - I inspected the index dir and the merge logs.
  - I triggered a couple of merge jobs with `curl http://localhost:6072/debug/merge` and confirmed that only 1 merge operation can run at a time.

stefanhengl committed:
merging: lock index dir during merge

stefanhengl committed:
hack: more verbose test output for RCA

stefanhengl committed:
mock zoekt-merge-index

CI doesn't have the executable zoekt-merge-index which is why the tests
failed previously. Now we mock it.

stefanhengl committed:
use new locking mechanism

stefanhengl commented:
@keegancsmith I rebased on master. This PR now uses the new locking mechanism. The core change is in doMerge(). The diff looks more complicated than it should so I recommend looking at the code directly.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/merge.go:
```diff
@@ -30,62 +30,90 @@ var metricShardMergingDuration = promauto.NewHistogramVec(prometheus.HistogramOp
+			if len(c.shards) <= 1 {
+				debug.Printf("could not find enough shards to build a compound shard\n")
+				next = false
+				return
```
minor: a way to make it easier not to miss setting next to false is to set it to false straight away and then on the happy path set next to true.

stefanhengl committed:
address PR comments

- elevate logs
- response on /debug/merge with "merging enqueued"
- simplify break condition
